157 Rank projects by user contributions

shahshlok opened the pull request:
## 📝 Description

This PR implements project ranking functionality and enhances the repository intelligence system for the ArtifactMiner application. It adds user-based project ranking, framework detection, database migrations, and comprehensive testing infrastructure.

The changes enable users to:
- Rank projects by contribution percentage using git shortlog analysis
- Detect project frameworks automatically
- Access user-specific repository intelligence
- Upload and analyze ZIP files through improved file handling

**Closes:** #157

---

## 🔧 Type of Change

- [x] ✨ New feature (non-breaking change that adds functionality)
- [x] ✅ Test added/updated
- [x] ♻️ Refactoring
- [] 📚 Documentation added/updated

---

## 🧪 Testing

Manual testing performed:
- [x] Project ranker correctly identifies user email addresses and calculates contribution percentages
- [x] ZIP file upload and extraction functionality working properly


Unit tests added:
- [x] test_project_ranker.py - Tests for contribution ranking logic

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [x] 📱 UI changes tested (TUI file browser and upload screens updated)

---

## 📸 Screenshots

<img width="1702" height="309" alt="Screenshot 2025-11-23 at 4 22 50 PM" src="https://github.com/user-attachments/assets/8b8380ab-63b2-47dd-bf78-510bded246da" />


---

## Summary of Changes

**New Files:**
- `src/artifactminer/helpers/project_ranker.py` - Project ranking engine

Whiteknight07 commented on tests/helpers/test_project_ranker.py:
No coverage for invalid/empty projects_dir returning [], directories without a .git folder being skipped, repositories with zero commits (should give 0.0 scores without divide-by-zero)

how about these test cases?

Whiteknight07 reviewed (changes requested):
Probably need more testing for `project_ranker.py`. What do you think?

shahshlok commented:
@Whiteknight07  I added the test cases that you requested and a few more. Here is a list:
I added the following test cases to handle various edge cases and ensure robustness:


- `test_rank_projects_invalid_path`: Verifies that a non‑existent path returns an empty list.
- `test_rank_projects_empty_directory`: Verifies that an empty directory returns an empty list.
- `test_rank_projects_skip_non_git`: Verifies that directories without a `.git` folder are correctly skipped.
- `test_rank_projects_zero_commits`: Verifies that repositories with 0 commits are handled without division‑by‑zero errors (returns a score of 0.0).
- `test_rank_projects_git_failure`: Mocks `subprocess.check_output` to raise `CalledProcessError` and ensures the function handles it gracefully (returns an empty list or skips the project).
- `test_rank_projects_malformed_output`: Mocks `subprocess.check_output` with unparseable lines to ensure the parser is robust.
- `test_rank_projects_case_insensitivity`: Verifies that user‑email matching is case‑insensitive.
- `test_rank_projects_no_user_commits`: Verifies correct handling when the user has 0 commits in a repository that has other commits.

Whiteknight07 reviewed (approved):
Great job on the new tests. Good to see that you did not have to refactor your `project_ranker.py` at all for this. 
Approved.

ahmadmemon reviewed (approved):
Everything makes sense and the test cases pass as well. Well done. Approved
